File: built-in plugins dsp/mb-mix-summing.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixSummingPlugin: PluginDefinition = { id: 'mb-mix-summing', slug: 'mb-mix-summing', name: 'MB Analog Summing', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'Analog summing bus emulation for warmth and width', author: 'Max Booster', parameters: [{ id: 'saturation', name: 'Saturation', type: 'float', defaultValue: 0.25, minValue: 0, maxValue: 1, automatable: true }, { id: 'width', name: 'Width', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'color', name: 'Color', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }, { id: 'output', name: 'Output', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { saturation: 0.25, width: 0.5, color: 0.4, output: 0.8 } };

export default MbMixSummingPlugin;

File: built-in plugins dsp/mb-mix-summing.cpp
/**
 * MB Analog Summing
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Analog summing bus emulation for warmth and width
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_SUMMING_H
#define MB_MIX_SUMMING_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixSumming : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-summing";
    static constexpr const char* PLUGIN_NAME    = "MB Analog Summing";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float saturation = 0.25f;  // range [0, 1]
    float width = 0.5f;  // range [0, 1]
    float color = 0.4f;  // range [0, 1]
    float output = 0.8f;  // range [0, 1]
    };

    MbMixSumming() = default;
    ~MbMixSumming() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.saturation = std::clamp(params.saturation, 0f, 1f);
        params.width = std::clamp(params.width, 0f, 1f);
        params.color = std::clamp(params.color, 0f, 1f);
        params.output = std::clamp(params.output, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Analog Summing
        return input;
    }
};

#endif // MB_MIX_SUMMING_H
